streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java

```diff
@@ -932,14 +937,59 @@ private ProcessorTopology build(final Set<String> nodeGroup) {
                 }
             }
         }
+        processorTopology = new ProcessorTopology(new ArrayList<>(processorMap.values()),
+                topicSourceMap,
+                topicSinkMap,
+                new ArrayList<>(stateStoreMap.values()),
+                new ArrayList<>(globalStateStores.values()),
+                storeToChangelogTopic,
+                repartitionTopics,
+                stateStoreToChangelogTopicOnlyForRestoring);
+        optimizeTopology();
+        return processorTopology;
+    }
+
+    /**
+     * Topology optimizer that avoids creating changelog topics for state stores that are directly piped to a sink topic
+     */
+    private void optimizeTopology() {
+        // Find processor nodes with changelog stores
+        List<ProcessorNode> nodesWithChangeLogState = new ArrayList<>();
+        for (String key : storeToChangelogTopic.keySet()) {
+            for (ProcessorNode processorNode: processorTopology.processors()) {
+                if (processorNode.stateStores != null && processorNode.stateStores.contains(key)) {
+                    nodesWithChangeLogState.add(processorNode);
+                }
+            }
+        }
 
-        return new ProcessorTopology(new ArrayList<>(processorMap.values()),
-                                     topicSourceMap,
-                                     topicSinkMap,
-                                     new ArrayList<>(stateStoreMap.values()),
-                                     new ArrayList<>(globalStateStores.values()),
-                                     storeToChangelogTopic,
-                                     repartitionTopics);
+        // If child nodes of nodesWithChangeLogState are of SinkNode type, then we can apply the optimization.
+        for (ProcessorNode processorNode: nodesWithChangeLogState) {
+            for (ProcessorNode child: (List<ProcessorNode>) processorNode.children()) {
+                if (child instanceof SinkNode) {
+                    for (String stateStore: (Set<String>) processorNode.stateStores) {
+                        stateStoreToChangelogTopicOnlyForRestoring.put(stateStore, storeToChangelogTopic.get(stateStore));
+                    }
+                    storeToChangelogTopic.keySet().removeAll(processorNode.stateStores);
```

Comment:
Can this be lifted out of the for loop by utilizing stateStoreToChangelogTopicOnlyForRestoring.keySet() ?

Reply:
Sorry but I don't get your point.

Reply:
I think one call of storeToChangelogTopic.keySet().removeAll() outside the loop should be equivalent to what you have now.

Reply:
Please specify exactly what you mean. I don't get it. This is general for all your comments: In theory, everything can be optimized for a specific JVM. The intuition here is that we will not have millions of operators in QEP. So for java, looping over 5 or 50 results a negligible difference. Therefore, I prefer readability over 0.93 nanosecond performance improvement, in logical query optimization part.

Reply:
My previous comments were only suggestions.

To me, expressing similar logic with one nested loop is more readable than writing two nested loops.
```
prefer readability over 0.93 nanosecond performance improvement
```
Not sure where 0.93 came from. IMHO if we know certain formation would gain better performance, we should use it.